bech32/app/Main.hs

```diff
@@ -0,0 +1,142 @@
+
+-- | Parse command line options and arguments
+parse :: IO Cmd
+parse = customExecParser (prefs showHelpOnEmpty) parser
+  where
+    parser :: ParserInfo Cmd
+    parser = info (helper <*> cmd) $ mconcat
+        [ progDesc "Simple utility for converting to/from bech32 byte strings."
+        ]
+
+    cmd :: Parser Cmd
+    cmd = Cmd <$> optional hrpArgument
+
+-- | Parse a 'HumanReadablePart' as an argument.
+hrpArgument :: Parser HumanReadablePart
+hrpArgument = argument (eitherReader reader) $ mempty
+    <> metavar "PREFIX"
+    <> help "A human-readable prefix (e.g. 'addr')."
+  where
+    reader :: String -> Either String HumanReadablePart
+    reader = left show . humanReadablePartFromText . T.pack
+
+-- | Run a Command in IO
+run :: Cmd -> IO ()
+run Cmd{prefix} = do
+    source <- T.decodeUtf8 . B8.filter (/= '\n') <$> B8.hGetContents stdin
+    case prefix of
+        Nothing  -> runDecode source
+        Just hrp -> runEncode hrp source
+  where
+    runDecode source =
+        case Bech32.decodeLenient source of
+          Left err ->
+              fail (show err)
+          Right (_, dataPart) ->
+              B8.putStrLn $ convertToBase BA.Base16 $ fromJust $
+                dataPartToBytes dataPart
+
+    runEncode hrp source = do
+        datapart <- either fail pure $
+            case detectEncoding (T.unpack source) of
+                Just Base16 -> do
+                    let fromBase16 = convertFromBase BA.Base16 . T.encodeUtf8
+                    dataPartFromBytes <$> fromBase16 source
+                Just Bech32 ->
+                    right snd $ left show $ Bech32.decodeLenient source
+                Just Base58 -> do
+                    let maybeToEither = maybe (Left "Invalid Base58-encoded string.") Right
+                    let fromBase58 = decodeBase58 bitcoinAlphabet . T.encodeUtf8
+                    dataPartFromBytes <$> maybeToEither (fromBase58 source)
+                Nothing ->
+                    fail "Unable to detect input encoding. Neither Base16, Bech32 nor Base58."
+        B8.putStrLn $ T.encodeUtf8 $ Bech32.encodeLenient hrp datapart
+
+data Encoding = Base16 | Bech32 | Base58 deriving (Show, Eq)
+
+-- | Try detecting the encoding of a given 'String'
+detectEncoding :: String -> Maybe Encoding
+detectEncoding str = isBase16 <|> isBech32  <|> isBase58
+  where
+    isBase16 = do
+        guard (all (`elem` "0123456789abcdef") (toLower <$> str))
+        guard (even (length str))
+        pure Base16
+
+    isBech32 = do
+        guard (not (null humanpart))
+        guard (all (\c -> ord c >= 33 && ord c <= 126) humanpart)
+        guard (length datapart >= 6)
+        guard (all (`elem` Bech32.dataCharList) datapart)
+        guard (all isUpper alpha || all isLower alpha)
+        pure Bech32
+      where
+        datapart  = reverse . takeWhile (/= '1') . reverse $ str
+        humanpart = takeWhile (/= '1') str
+        alpha = filter isLetter str
```

Comment:
**Suggestion**: reuse the existing Bech32 parser.

For example:
```hs
    isBech32 = str    
        & T.pack      
        & Bech32.decodeLenient
        & eitherToMaybe       
        & fmap (const Bech32)
```
(Where `eitherToMaybe` is from `Data.Either.Extra`, which we already depend upon. :+1: )

By doing this, we can also avoid hard-coding several magic constants (for example, `6`, `33`, `126`, and so on).

Reply:
Not really. The goal is to "detect" whether a given string _looks like_ a bech32 string, so that we can actually try decoding it. The string might be invalid, but then we can show some adequate error message. If we were to fully decode it and try the next parser in case of failure, then we may end up showing error messages for the next parser that are quite unrelated to the actual string. 

There's a bit of duplication indeed, but that's a small price to pay for a nice UX. And it's not like these constants are changing anytime soon ;)

Reply:
> Not really. The goal is to "detect" whether a given string _looks like_ a bech32 string

In that case, I had mistaken the intent of the code. The name `isBech32` seems to imply that the function really does check the string to see if it is a bech32 string. Perhaps we could add a comment here to clarify the intent.

In any case, I suspect we should add some tests here.

For example, the following string is accepted by all of the checking functions (`bech32`, `base64`, and `base58`):

```hs
>>> isBase58 "afafaf"
Base58
>>> isBech32 "afafaf"
Bech32
>>> isBase16 "afafaf"
Base16
```

We could fix `isBech32` by checking that the string has at least a single instance of `Bech32.separatorChar`.

As for the constant values, I would still be in favour of using named constants and functions here. For example:

```patch
- guard (all (\c -> ord c >= 33 && ord c <= 126) humanpart)
+ guard (all Bech32.humanReadableCharIsValid humanpart)
```

And:

```patch
- guard (length datapart >= 6)
+ guard (length datapart >= Bech32.checksumLength)
```

The justification here is to make the code more readable to someone who is not the original author, or even to our future selves!